libraries/AP_Proximity/AP_Proximity_RPLidarA2.cpp

```diff
@@ -101,260 +107,236 @@ float AP_Proximity_RPLidarA2::distance_min() const
-            static_assert(sizeof(payload.sensor_scan) == 5, "Unexpected payload.sensor_scan data structure size");
-            _response_type = ResponseType_SCAN;
-            Debug(2, "Measurement response detected");
-            _last_distance_received_ms = AP_HAL::millis();
-            _rp_state = rp_measurements;
+    Debug(2, "             CURRENT STATE: %u ", (unsigned)_state);
+    const uint32_t nbytes = _uart->available();
+    if (nbytes == 0) {
+        return;
+    }
+    const uint32_t bytes_to_read = MIN(nbytes, sizeof(_payload)-_byte_count);
+    if (bytes_to_read == 0) {
+        INTERNAL_ERROR(AP_InternalError::error_t::flow_of_control);
+        reset();
+        return;
+    }
+    const uint32_t bytes_read = _uart->read(&_payload[_byte_count], bytes_to_read);
+    if (bytes_read == 0) {
+        // this is bad; we were told there were bytes available
+        INTERNAL_ERROR(AP_InternalError::error_t::flow_of_control);
+        reset();
+        return;
+    }
+    _byte_count += bytes_read;
+
+    const uint32_t start_ms = AP_HAL::millis();
+    while (_byte_count) {
+        if (AP_HAL::millis() - start_ms > 300) {
+            // it should never take us 300ms to process ~128 bytes
+            // (sizeof(_payload)) (even under Valgrind)
+            INTERNAL_ERROR(AP_InternalError::error_t::flow_of_control);
+            _uart = nullptr;
+            return;
         }
-        if (_descriptor[2] == 0x03 && _descriptor[3] == 0x00 && _descriptor[4] == 0x00 && _descriptor[5] == 0x00 && _descriptor[6] == 0x06) {
-            // payload is health data
-            _payload_length = sizeof(payload.sensor_health);
-            static_assert(sizeof(payload.sensor_health) == 3, "Unexpected payload.sensor_health data structure size");
-            _response_type = ResponseType_Health;
-            _last_distance_received_ms = AP_HAL::millis();
-            _rp_state= rp_health;
+
+        switch(_state){
+        case State::RESET: {
+            // looking for 0x52 at start of buffer; the 62 following
+            // bytes are "information"
+            if (!make_first_byte_in_payload('R')) { // that's 'R' as in RPiLidar
+                return;
+            }
+            if (_byte_count < 63) {
+                return;
+            }
+#if RP_DEBUG_LEVEL
+            // optionally spit out via mavlink the 63-bytes of cruft
+            // that is spat out on device reset
+            Debug(1, "Got RPLidar Information");
+            char xbuffer[64]{};
+            memcpy((void*)xbuffer, (void*)&_payload.information, 63);
+            gcs().send_text(MAV_SEVERITY_INFO, "RPLidar: (%s)", xbuffer);
+#endif
+            // 63 is the magic number of bytes in the spewed-out
+            // reset data ... so now we'll just drop that stuff on
+            // the floor.
+            consume_bytes(63);
+            send_scan_mode_request();
+//            send_request_for_health();
```

Comment:
I guess we can remove this commented out line..

Reply:
Done